### PR TITLE
Complete array support

### LINKED ISSUE #194: Support for arrays
Most of the support for WinRT arrays is already in place, but a bit is missing and a lot of testing is still required to validate it all.

WinRT doesn't have an array type as such. Arrays only occur as parameters to WinRT APIs. That is why `winrt::Array<T>` does not implement the `RuntimeType` trait but is constrained to only allow arrays of `RuntimeType`.

There are three different array parameter patterns. 

1. An **input** array parameter where the caller provides a pointer to the array as well as the number of elements in the array that the callee may **read** from. Storage is provided by the caller (stack or heap).

2. An **output** array parameter where the caller provides a pointer to the array as well as the number of elements in the array that the callee may **write** to. Storage is provided by the caller (stack or heap).

3. An **output** array parameter where the callee returns a pointer allocated using `CoTaskMemAlloc` as well as the number of elements in the array and the caller assumes ownership of the storage and must ultimately call `CoTaskMemFree` to free the storage, but only after dropping the individual elements in the array. 

The `winrt::Array<T>` type needs to be completed to properly handle creating (allocating) and dropping arrays for (3) above. (1) and (2) should be handled directly by Rust slices (and is mostly already working I believe).

There are a few more subtleties to arrays particularly around the ABI handling, but that's the gist of it.

Comprehensive tests are needed for all three patterns.

### PR DESCRIPTION
**Pull request by kennykerr**
Fixes #194

Still needs more comprehensive testing #213, but this covers basic array support.

### CODE REVIEW

File: src/array.rs [hunk 1]
```diff
@@ -41,8 +53,30 @@ impl<T: RuntimeType> Array<T> {
...
 
-impl<T> Drop for Array<T> {
+impl<T: RuntimeType> std::ops::Deref for Array<T> {
+    type Target = [T];
+
+    fn deref(&self) -> &[T] {
+        if self.data.is_null() {
+            return &[];
+        }
+
```
Comment (rylev) [inline src/array.rs:63]: Just for some sanity checking might be useful to have a `debug_assert!` here that `self.len > 0`.

File: src/array.rs [hunk 2]
```diff
@@ -41,8 +53,30 @@ impl<T: RuntimeType> Array<T> {
...
+        unsafe { std::slice::from_raw_parts_mut(self.data, self.len as usize) }
+    }
+}
+
+impl<T: RuntimeType> Drop for Array<T> {
     fn drop(&mut self) {
-        // TODO: CoTaskMemFree
+        self.clear();
     }
 }
```
Comment (rylev) [inline src/array.rs:82]: Can we add some unit tests for Array?

File: src/array.rs [hunk 3]
```diff
@@ -16,19 +16,31 @@ impl<T> Default for Array<T> {
...
+
+    /// Clears the contents of the array.
+    pub fn clear(&mut self) {
+        if self.is_empty() {
+            return;
+        }
+
+        unsafe {
+            std::ptr::drop_in_place(&mut self[..]);
+            CoTaskMemFree(self.data as _);
```
Comment (rylev) [inline src/array.rs:42]: We need to null the pointer out, otherwise the following is UB:

```rust
my_array.clear();
let slice: &[T] = &*my_array;
```

File: src/array.rs [hunk 4]
```diff
@@ -16,19 +16,31 @@ impl<T> Default for Array<T> {
...
+    }
+
+    /// Clears the contents of the array.
+    pub fn clear(&mut self) {
+        if self.is_empty() {
+            return;
+        }
+
+        unsafe {
+            std::ptr::drop_in_place(&mut self[..]);
```
Comment (rylev) [inline src/array.rs:41]: I think this is not fully correct. If we call `clear` and one of the values in the array panics during its drop, we'll unwind the stack and call `clear` again in the drop impl of Array leading us right back here to the where the array elements drop impls will be called again (double free). 

I think we have to first swap the pointers and then drop. If one of the elements panics, we may leak memory, but we won't double free.

File: src/array.rs (hunk 1 above)
Comment (kennykerr) [reply src/array.rs:69]: Switched to `is_empty` for consistency.

File: src/array.rs (hunk 2 above)
Comment (kennykerr) [reply src/array.rs:88]: I have some basic tests that cover the various issues here in the tests/boxing.rs - its difficult to cover all the edge cases without #213 - that's up next.

File: src/array.rs (hunk 3 above)
Comment (kennykerr) [reply src/array.rs]: Fixed.

File: src/array.rs (hunk 4 above)
Comment (kennykerr) [reply src/array.rs]: Fixed.

File: src/array.rs [hunk 5]
```diff
@@ -16,19 +16,37 @@ impl<T> Default for Array<T> {
...
+    pub fn clear(&mut self) {
+        if self.is_empty() {
+            return;
+        }
+
+        let mut data = std::ptr::null_mut();
+        let mut len = 0;
+
+        std::mem::swap(&mut data, &mut self.data);
+        std::mem::swap(&mut len, &mut self.len);
```
Comment (rylev) [inline src/array.rs:44]: This isn’t really necessary. `self.len = 0` is also fine
